File: sdks/typescript/src/services/MemoryHelpers.ts
/**
 * Memory Helper Methods (SDK Sugar)
 *
 * High-level, opinionated helpers for common memory operations.
 * These methods use StandardContexts and provide agent-friendly abstractions.
 *
 * Philosophy: Guide agents toward best practices without enforcing restrictions.
 */

import type { MemoryService } from './MemoryService';
import { StandardContexts } from '../constants/StandardContexts';

export interface UserPreference {
  key: string;
  value: unknown;
  userId?: string;
  timestamp?: number;
}

export interface ErrorFix {
  error: string;
  solution: string;
  context?: string;
  timestamp?: number;
}

export interface Pattern {
  pattern: string;
  success: boolean;
  domain?: string;
  confidence?: number;
  timestamp?: number;
}

export interface ConversationSummary {
  summary: string;
  turns: number;
  startTime: number;
  endTime?: number;
  participants?: string[];
}

export interface ToolConfig {
  toolName: string;
  config: Record<string, unknown>;
  version?: string;
  timestamp?: number;
}

export interface DomainHeuristic {
  domain: string;
  heuristic: string;
  confidence: number;
  evidence?: string;
  timestamp?: number;
}

export interface OptimizationInsight {
  operation: string;
  improvement: string;
  metrics?: Record<string, number>;
  timestamp?: number;
}

export interface Memory {
  storageKey: string;
  data: Record<string, unknown>;
  storageTier: string;
  metadata?: Record<string, unknown>;
  receiptId: string;
}

export class MemoryHelpers {
  constructor(private readonly memoryService: MemoryService) {}

  /**
   * Remember a user preference
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberPreference({
   *   key: 'responseStyle',
   *   value: 'concise',
   *   userId: 'user_123'
   * });
   * ```
   */
  async rememberPreference(pref: UserPreference): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        key: pref.key,
        value: pref.value,
        userId: pref.userId,
        timestamp: pref.timestamp || Date.now(),
      },
      storageTier: 'hot',
      context: StandardContexts.USER_PREFERENCE,
      tags: ['preference', pref.key],
      metadata: {
        preferenceKey: pref.key,
        userId: pref.userId,
      },
    });

    return result.storageKey;
  }

  /**
   * Remember an error fix
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberFix({
   *   error: 'TypeError: undefined is not a function',
   *   solution: 'Added optional chaining operator',
   *   context: 'payment-handler'
   * });
   * ```
   */
  async rememberFix(fix: ErrorFix): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        error: fix.error,
        solution: fix.solution,
        context: fix.context,
        timestamp: fix.timestamp || Date.now(),
      },
      storageTier: 'warm',
      context: StandardContexts.ERROR_FIX,
      tags: ['error', 'fix'],
      metadata: {
        errorType: this.extractErrorType(fix.error),
      },
    });

    return result.storageKey;
  }

  /**
   * Remember a pattern (successful or failed)
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberPattern({
   *   pattern: 'Use exponential backoff for API retries',
   *   success: true,
   *   domain: 'api-resilience',
   *   confidence: 0.95
   * });
   * ```
   */
  async rememberPattern(pattern: Pattern): Promise<string> {
    const context = pattern.success
      ? StandardContexts.SUCCESSFUL_PATTERN
      : StandardContexts.FAILED_APPROACH;

    const result = await this.memoryService.store({
      data: {
        pattern: pattern.pattern,
        success: pattern.success,
        domain: pattern.domain,
        confidence: pattern.confidence,
        timestamp: pattern.timestamp || Date.now(),
      },
      storageTier: 'warm',
      context,
      tags: ['pattern', pattern.success ? 'success' : 'failure', pattern.domain].filter(Boolean) as string[],
      metadata: {
        domain: pattern.domain,
        confidence: pattern.confidence,
      },
    });

    return result.storageKey;
  }

  /**
   * Remember a conversation summary
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberConversation({
   *   summary: 'User asked about restaurant recommendations, I suggested 3 options',
   *   turns: 5,
   *   startTime: Date.now() - 300000,
   *   endTime: Date.now()
   * });
   * ```
   */
  async rememberConversation(conv: ConversationSummary): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        summary: conv.summary,
        turns: conv.turns,
        startTime: conv.startTime,
        endTime: conv.endTime || Date.now(),
        participants: conv.participants,
      },
      storageTier: 'hot',
      context: StandardContexts.CONVERSATION_SUMMARY,
      tags: ['conversation', `turns:${conv.turns}`],
      metadata: {
        turnCount: conv.turns,
        duration: (conv.endTime || Date.now()) - conv.startTime,
      },
    });

    return result.storageKey;
  }

  /**
   * Remember a tool configuration
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberToolConfig({
   *   toolName: 'weatherAPI',
   *   config: { apiKey: 'xxx', units: 'metric' },
   *   version: '2.0'
   * });
   * ```
   */
  async rememberToolConfig(tool: ToolConfig): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        toolName: tool.toolName,
        config: tool.config,
        version: tool.version,
        timestamp: tool.timestamp || Date.now(),
      },
      storageTier: 'warm',
      context: StandardContexts.TOOL_CONFIG,
      tags: ['tool', tool.toolName],
      metadata: {
        toolName: tool.toolName,
        version: tool.version,
      },
    });

    return result.storageKey;
  }

  /**
   * Remember a domain-specific heuristic
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberHeuristic({
   *   domain: 'code-review',
   *   heuristic: 'Functions over 50 lines should be refactored',
   *   confidence: 0.85,
   *   evidence: 'Based on 100+ code reviews'
   * });
   * ```
   */
  async rememberHeuristic(heuristic: DomainHeuristic): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        domain: heuristic.domain,
        heuristic: heuristic.heuristic,
        confidence: heuristic.confidence,
        evidence: heuristic.evidence,
        timestamp: heuristic.timestamp || Date.now(),
      },
      storageTier: 'cold',
      context: StandardContexts.DOMAIN_HEURISTIC,
      tags: ['heuristic', heuristic.domain],
      metadata: {
        domain: heuristic.domain,
        confidence: heuristic.confidence,
      },
    });

    return result.storageKey;
  }

  /**
   * Remember an optimization insight
   *
   * @example
   * ```typescript
   * const storageKey = await client.memory.rememberOptimization({
   *   operation: 'database-query',
   *   improvement: 'Added index on user_id column',
   *   metrics: { before: 2500, after: 150, unit: 'ms' }
   * });
   * ```
   */
  async rememberOptimization(insight: OptimizationInsight): Promise<string> {
    const result = await this.memoryService.store({
      data: {
        operation: insight.operation,
        improvement: insight.improvement,
        metrics: insight.metrics,
        timestamp: insight.timestamp || Date.now(),
      },
      storageTier: 'warm',
      context: StandardContexts.OPTIMIZATION_INSIGHT,
      tags: ['optimization', insight.operation],
      metadata: {
        operation: insight.operation,
        metrics: insight.metrics,
      },
    });

    return result.storageKey;
  }

  /**
   * Recall all user preferences
   *
   * @example
   * ```typescript
   * const prefs = await client.memory.recallPreferences({ limit: 50 });
   * for (const pref of prefs) {
   *   console.log(pref.data.key, pref.data.value);
   * }
   * ```
   */
  async recallPreferences(opts?: { limit?: number }): Promise<Memory[]> {
    const result = await this.memoryService.list({
      context: StandardContexts.USER_PREFERENCE,
      limit: opts?.limit || 50,
    });

    // List returns metadata, not decrypted data - retrieve each memory
    const memories: Memory[] = [];
    for (const item of result.memories || []) {
      try {
        const retrieved = await this.memoryService.retrieve({ storageKey: item.storage_key });
        memories.push({
          storageKey: retrieved.storageKey,
          data: retrieved.data as Record<string, unknown>,
          storageTier: retrieved.storageTier,
          metadata: retrieved.metadata,
          receiptId: retrieved.receiptId,
        });
      } catch {
        // Skip failed retrievals
      }
    }
    return memories;
  }

  /**
   * Recall error fixes
   *
   * @example
   * ```typescript
   * const fixes = await client.memory.recallFixes({ domain: 'api-errors' });
   * ```
   */
  async recallFixes(opts?: { domain?: string; limit?: number }): Promise<Memory[]> {
    const result = await this.memoryService.list({
      context: StandardContexts.ERROR_FIX,
      limit: opts?.limit || 50,
    });

    const memories: Memory[] = [];
    for (const item of result.memories || []) {
      try {
        const retrieved = await this.memoryService.retrieve({ storageKey: item.storage_key });
        // Filter by domain if specified
        if (opts?.domain && (retrieved.data as any)?.context !== opts.domain) {
          continue;
        }
        memories.push({
          storageKey: retrieved.storageKey,
          data: retrieved.data as Record<string, unknown>,
          storageTier: retrieved.storageTier,
          metadata: retrieved.metadata,
          receiptId: retrieved.receiptId,
        });
      } catch {
        // Skip failed retrievals
      }
    }
    return memories;
  }

  /**
   * Recall patterns
   *
   * @example
   * ```typescript
   * const successPatterns = await client.memory.recallPatterns({ success: true });
   * ```
   */
  async recallPatterns(opts?: { success?: boolean; limit?: number }): Promise<Memory[]> {
    // Query both success and failure contexts based on filter
    const context = opts?.success === true
      ? StandardContexts.SUCCESSFUL_PATTERN
      : opts?.success === false
        ? StandardContexts.FAILED_APPROACH
        : undefined;

    const result = await this.memoryService.list({
      context,
      limit: opts?.limit || 50,
    });

    const memories: Memory[] = [];
    for (const item of result.memories || []) {
      try {
        const retrieved = await this.memoryService.retrieve({ storageKey: item.storage_key });
        memories.push({
          storageKey: retrieved.storageKey,
          data: retrieved.data as Record<string, unknown>,
          storageTier: retrieved.storageTier,
          metadata: retrieved.metadata,
          receiptId: retrieved.receiptId,
        });
      } catch {
        // Skip failed retrievals
      }
    }
    return memories;
  }

  /**
   * Extract error type from error message
   * @private
   */
  private extractErrorType(error: string): string {
    const match = error.match(/^(\w+Error):/);
    return match ? match[1] : 'UnknownError';
  }
}

File: sdks/python/xache/services/memory_helpers.py
"""
Memory Helpers — High-level convenience methods for common memory operations.

Uses StandardContexts for interoperable memory patterns.
"""

from typing import Any, Dict, List, Optional

from ..constants.standard_contexts import StandardContexts


class MemoryHelpers:
    """
    Convenience methods for common memory store/retrieve patterns.

    Example:
        ```python
        helpers = MemoryHelpers(client)

        # Store a user preference
        key = await helpers.remember_preference(
            "dark_mode", True,
            subject=subject_ctx,
        )

        # Recall all preferences
        prefs = await helpers.recall_preferences(subject=subject_ctx)
        ```
    """

    def __init__(self, client: Any):
        self.client = client

    # ========== Remember (Store) Methods ==========

    async def remember_preference(
        self,
        key: str,
        value: Any,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a user preference (hot storage)."""
        result = await self.client.memory.store(
            data={"key": key, "value": value},
            storage_tier="hot",
            context=StandardContexts.USER_PREFERENCE,
            tags=["preference", key],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_fix(
        self,
        error: str,
        solution: str,
        domain: Optional[str] = None,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store an error-to-solution mapping (warm storage)."""
        data: Dict[str, Any] = {"error": error, "solution": solution}
        if domain:
            data["domain"] = domain
        tags = ["error-fix"]
        if domain:
            tags.append(f"domain:{domain}")

        result = await self.client.memory.store(
            data=data,
            storage_tier="warm",
            context=StandardContexts.ERROR_FIX,
            tags=tags,
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_pattern(
        self,
        pattern: str,
        success: bool,
        domain: Optional[str] = None,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a successful or failed pattern (warm storage)."""
        ctx = StandardContexts.SUCCESSFUL_PATTERN if success else StandardContexts.FAILED_APPROACH
        data: Dict[str, Any] = {"pattern": pattern, "success": success}
        if domain:
            data["domain"] = domain

        result = await self.client.memory.store(
            data=data,
            storage_tier="warm",
            context=ctx,
            tags=["pattern", "success" if success else "failure"],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_conversation(
        self,
        summary: str,
        key_points: Optional[List[str]] = None,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a conversation summary (hot storage)."""
        data: Dict[str, Any] = {"summary": summary}
        if key_points:
            data["keyPoints"] = key_points

        result = await self.client.memory.store(
            data=data,
            storage_tier="hot",
            context=StandardContexts.CONVERSATION_SUMMARY,
            tags=["conversation"],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_tool_config(
        self,
        tool: str,
        config: Dict[str, Any],
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a tool configuration (warm storage)."""
        result = await self.client.memory.store(
            data={"tool": tool, "config": config},
            storage_tier="warm",
            context=StandardContexts.TOOL_CONFIG,
            tags=["tool-config", tool],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_heuristic(
        self,
        insight: str,
        domain: str,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store a domain-specific heuristic (cold storage)."""
        result = await self.client.memory.store(
            data={"insight": insight, "domain": domain},
            storage_tier="cold",
            context=StandardContexts.DOMAIN_HEURISTIC,
            tags=["heuristic", f"domain:{domain}"],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    async def remember_optimization(
        self,
        insight: str,
        category: Optional[str] = None,
        subject: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Store an optimization insight (warm storage)."""
        data: Dict[str, Any] = {"insight": insight}
        if category:
            data["category"] = category

        result = await self.client.memory.store(
            data=data,
            storage_tier="warm",
            context=StandardContexts.OPTIMIZATION_INSIGHT,
            tags=["optimization"],
            **(self._subject_kwargs(subject)),
        )
        return result.memory_id

    # ========== Recall (Retrieve) Methods ==========

    async def recall_preferences(
        self,
        subject: Optional[Dict[str, Any]] = None,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Retrieve all user preferences."""
        result = await self.client.memory.list(
            context=StandardContexts.USER_PREFERENCE,
            limit=limit,
            **(self._subject_kwargs(subject)),
        )
        return [{"storage_key": m.storage_key, "tags": m.tags, "metadata": m.metadata}
                for m in result.memories]

    async def recall_fixes(
        self,
        domain: Optional[str] = None,
        subject: Optional[Dict[str, Any]] = None,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Retrieve error fixes, optionally filtered by domain."""
        result = await self.client.memory.list(
            context=StandardContexts.ERROR_FIX,
            limit=limit,
            **(self._subject_kwargs(subject)),
        )
        memories = result.memories
        if domain:
            memories = [m for m in memories if f"domain:{domain}" in (m.tags or [])]
        return [{"storage_key": m.storage_key, "tags": m.tags, "metadata": m.metadata}
                for m in memories]

    async def recall_patterns(
        self,
        success_only: Optional[bool] = None,
        subject: Optional[Dict[str, Any]] = None,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Retrieve patterns, optionally filtered by success/failure."""
        if success_only is True:
            ctx = StandardContexts.SUCCESSFUL_PATTERN
        elif success_only is False:
            ctx = StandardContexts.FAILED_APPROACH
        else:
            # Return both — need two calls
            successes = await self.client.memory.list(
                context=StandardContexts.SUCCESSFUL_PATTERN,
                limit=limit,
                **(self._subject_kwargs(subject)),
            )
            failures = await self.client.memory.list(
                context=StandardContexts.FAILED_APPROACH,
                limit=limit,
                **(self._subject_kwargs(subject)),
            )
            all_memories = successes.memories + failures.memories
            return [{"storage_key": m.storage_key, "tags": m.tags, "metadata": m.metadata}
                    for m in all_memories[:limit]]

        result = await self.client.memory.list(
            context=ctx,
            limit=limit,
            **(self._subject_kwargs(subject)),
        )
        return [{"storage_key": m.storage_key, "tags": m.tags, "metadata": m.metadata}
                for m in result.memories]

    # ========== Helpers ==========

    @staticmethod
    def _subject_kwargs(subject: Optional[Dict[str, Any]]) -> Dict[str, Any]:
        """Convert subject context dict to store/list kwargs."""
        if not subject:
            return {}
        kwargs: Dict[str, Any] = {}
        if subject.get("subject_id"):
            kwargs["subject_id"] = subject["subject_id"]
        if subject.get("scope"):
            kwargs["scope"] = subject["scope"]
        if subject.get("segment_id"):
            kwargs["segment_id"] = subject["segment_id"]
        if subject.get("tenant_id"):
            kwargs["tenant_id"] = subject["tenant_id"]
        return kwargs
